CONVENTIONS.md

```diff
@@ -0,0 +1,75 @@
+Please refer to this document whenever you do not know how to format or structure a name.
+
+Getting Started
+---------------
+ - **If you do not know what to name something then *do not name it*. Let someone else do so.**
+ - Names should reflect intent, *not* implementation. Check usages rather than just looking at the definition.
+ - Names should be concise and memorable. If information can be inferred from context (type, parameters etc.),
+   do not include it in the name. This especially applies to fields and methods in a class, where you do not
+   need to have the class name in the member's name.
+ - Prefer clarification in the documentation comment over verbose names. Conversely, do not give fields,
+   methods, or method parameters useless comments.
+
+General
+-------
+ - All names use US spelling.
+ - Acronyms should generally be written all-lowercase, while normal camel case rules still apply.
+   For instance, "identifier" becomes `id` as standalone word or at the beginning of a camel case phrase and
+   `Id` when used within a phrase.<br>
+   There are various exceptions to this rule which apply only when the word does not start with them:
+     - axis-aligned bounding boxes: `AxisAlignedBB` → `AABB`
+     - named binary tags: `NBT`
+     - red, green, blue (and alpha) color components → `RGB(A)`
+
+Type Names
+----------
+ - New classes, interfaces etc. should follow a suffix-based naming scheme.
+   Specifically, this means for any name, `NameType` would be the correct name, where `Type` should be
+   derived from the supertypes.
+ - Abstract base classes should be prefixed with `Abstract`.
+ - Interfaces are prefixed with the letter `I`.
+ - Enums should be named directly after the things they are enumerating, in singular form.
+   There is no `Enum` prefix/suffix.
+
+**Note:** Long-established names where prefixes are used (subclasses `Block`, `Item` etc.)
+do *not* fall under the suffix convention.
+
+### Examples
+ - For a new shelf block with an inventory, appropriate names would be `BlockShelf`, `ShelfTileEntity` and
+   `ShelfContainer`.
+
+Field Names
+-----------
+ - `boolean` type fields should *not* start with an *is* prefix unless there is a convincing reason to do so.
+ - No prefixes like `the` or suffixes like `Obj`.
+
+Method Names
+------------
+ - Method names should always start with a verb. Standard conventions for setters and getters apply.
+ - Methods returning a `boolean` should have a name beginning with an appropriate prefix from the following list:
+     - `can` 
+     - `does`
+     - `has`
+     - `is`
+     - `should`
+     - `was`
+ - 'Event handling' methods should follow an `on<Noun>` theme where the noun is the event's name
+   (with redundancies removed).
+ - Methods which read or write NBT should be prefixed `read` and `write` respectively.
+ - Methods which deseralize or serialize JSON should have a `deserialize` or `serialize` prefix accordingly. 
```

Comment:
Maybe clarify the difference between `read`/`write` and `deserialize`/`serialize`:
The "read"/"write" prefixes should be used when no new instances are created:
 - `something.read(NBTTagCompound tag)`
 - `something.write(PacketBuf buffer)`

The "serialize"/"deserialize" prefixes should be used when a new instance (of either the serialized or deserialized object) is created:
 - `Something something = somethingSerializer.deserialize(json)`
 - `JsonObject json = something.serializeToJson()`
 - `JsonObject json = somethingSerializer.serialize(something)`

This seems to be the convention we're using and is also consistent with how java and gson use the word "deserialize".

Reply:
`read` and `write` should be used for NBT and network methods, while `serialize` and `deserialize` should be used for JSON methods.

Reply:
> read and write should be used for NBT and network methods, while serialize and deserialize should be used for JSON methods.

Json uses the words `serialize`/`deserialize` because unlike the NBT and network methods, it doesn't just load data from a buffer or NBT tag into an existing instance, but instead creates a new instance of the object from the json.

read = load data into an *existing* object
write = save the object to a buffer or NBT tag *passed as a parameter*
deserialize = *create a new object* from a serialized form (json in this case)
serialize = *create* a serialized form of the object

This works with the definition of [serialization](https://en.wikipedia.org/wiki/Serialization), which is the conversion between serialized and deserialized objects, not just loading data into an object from a tag/stream.